### PR TITLE
PMM-5988 Implement API needed for progress bar

### PR DESCRIPTION
**Pull request by jprukner**
We can track progress of db cluster deployment thanks to new fields
TotalSteps and FinishedSteps in controllerv1beta1.RunningOperation.

### CODE REVIEW

File: service/k8sclient/k8client.go [hunk 1]
```diff
@@ -143,25 +143,37 @@ type PSMDBParams struct {
 	Replicaset       *Replicaset
 }
 
+type appStatus struct {
+	size  int32
+	ready int32
+}
+
+// DetailedState contains containers' status.
```
Comment (jprukner) [inline service/k8sclient/k8client.go:151]: pods

File: service/k8sclient/k8client.go [hunk 2]
```diff
@@ -143,25 +143,37 @@ type PSMDBParams struct {
...
+type DetailedState struct {
```
Comment (jprukner) [inline service/k8sclient/k8client.go:152]: mayby just 
```
type DetailedState []appstatus
```
would be enough

File: service/k8sclient/k8client.go [hunk 3]
```diff
@@ -502,6 +528,14 @@ func (c *K8Client) getPerconaXtraDBClusters(ctx context.Context) ([]XtraDBCluste
 				ComputeResources: c.getComputeResources(cluster.Spec.ProxySQL.Resources),
 			},
 			Pause: cluster.Spec.Pause,
+			DetailedState: &DetailedState{
+				status: []appStatus{
+					appStatus{cluster.Status.PMM.Ready, cluster.Status.PMM.Size},
```
Comment (BupycHuk) [inline service/k8sclient/k8client.go:533]: Make ensure that we have these fields and we won't get panic

File: service/k8sclient/k8client_test.go [hunk 4]
```diff
@@ -126,12 +146,31 @@ func TestK8Client(t *testing.T) {
...
+					cluster = c
+					return
+				}
+			}
+			return
+		}
+
+		t.Run("All pods are ready", func(t *testing.T) {
+			cluster := getCluster(name)
+			assert.Equal(t, cluster.DetailedState.CountReadyPods(), 6)
```
Comment (jprukner) [inline service/k8sclient/k8client_test.go:170]: int32
reverse order of args - expected, actual

File: service/k8sclient/k8client.go [hunk 5]
```diff
@@ -803,6 +838,12 @@ func (c *K8Client) getPSMDBClusters(ctx context.Context) ([]PSMDBCluster, error)
 		if message == "" && len(conditions) > 0 {
 			message = conditions[len(conditions)-1].Message
 		}
+
+		status := make([]appStatus, 0, len(cluster.Status.Replsets))
+		for _, rs := range cluster.Status.Replsets {
+			status = append(status, appStatus{rs.Size, rs.Ready})
```
Comment (BupycHuk) [inline service/k8sclient/k8client.go:844]: We also have mongos and config pods, let's include them too

File: service/k8sclient/k8client_test.go [hunk 6]
```diff
@@ -126,12 +146,31 @@ func TestK8Client(t *testing.T) {
 		})
 		require.NoError(t, err)
 
-		l.Info("PSMDB Cluster is created")
-
 		assertListPSMDBCluster(t, ctx, client, name, func(cluster *PSMDBCluster) bool {
 			return cluster != nil && cluster.State == ClusterStateReady
 		})
 
+		getCluster := func(name string) (cluster PSMDBCluster) {
```
Comment (BupycHuk) [inline service/k8sclient/k8client_test.go:153]: Let's move this function out of current test and reuse it in assertListxxxCluster functions

File: service/k8sclient/k8client.go (hunk 3 above)
Comment (jprukner) [reply service/k8sclient/k8client.go:533]: Well, those fields are not pointers so the worst that can happen is we get zero value for Ready and Size, right?

File: service/k8sclient/k8client_test.go (hunk 6 above)
Comment (jprukner) [reply service/k8sclient/k8client_test.go:153]: Ok, I moved it and reused it.

In some future sprint, we should refactor at least tests so we don't duplicate code so much.

File: service/k8sclient/k8client.go (hunk 5 above)
Comment (jprukner) [reply service/k8sclient/k8client.go:840]: Included them by creating new field cluster.Status.Mongos.

**Comment by AlekSi** [comment]
There are conflicts

**Comment by jprukner** [comment]
> There are conflicts

Rebased on latest main.

**Comment by AlekSi** [comment]
CI fails due to what looks like mis-merge. Please compile code locally :)

**Review by BupycHuk — CHANGES_REQUESTED** [review]

File: service/k8sclient/k8sclient_test.go [hunk 7]
```diff
@@ -176,50 +191,65 @@ func TestK8Client(t *testing.T) {
...
-		l.Debug(clusters)
-
-		var cluster *PSMDBCluster
-		for _, c := range clusters {
-			c := c
-			if c.Name == name {
-				cluster = &c
-				break
-			}
+		time.Sleep(5 * time.Second)
```
[the remark this answers is not in the document]
Comment (jprukner) [reply service/k8sclient/k8sclient_test.go]: Done

File: service/k8sclient/k8sclient_test.go [hunk 8]
```diff
@@ -128,18 +136,25 @@ func TestK8Client(t *testing.T) {
...
-		assertListPSMDBCluster(t, ctx, client, name, func(cluster *PSMDBCluster) bool {
+		assertListPSMDBCluster(ctx, t, client, name, func(cluster *PSMDBCluster) bool {
 			return cluster != nil && cluster.State == ClusterStateReady
 		})
 
+		t.Run("All pods are ready", func(t *testing.T) {
+			cluster, err := getPSMDBCluster(ctx, client, name)
+			require.NoError(t, err)
+			assert.Equal(t, cluster.DetailedState.CountAllPods(), cluster.DetailedState.CountReadyPods())
+			assert.Greater(t, cluster.DetailedState.CountAllPods(), int32(0))
```
[the remark this answers is not in the document]
Comment (jprukner) [reply service/k8sclient/k8sclient_test.go:178]: Done

File: service/k8sclient/k8sclient_test.go [hunk 9]
```diff
@@ -69,17 +74,39 @@ func TestK8Client(t *testing.T) {
...
+		assertListXtraDBCluster(ctx, t, client, name, func(cluster *XtraDBCluster) bool {
 			return cluster != nil && cluster.State == ClusterStateReady
 		})
 
+		t.Run("All pods are ready", func(t *testing.T) {
+			var xtradbCluster pxc.PerconaXtraDBCluster
+			err := client.kubeCtl.Get(ctx, string(perconaXtraDBClusterKind), name, &xtradbCluster)
+			require.NoError(t, err)
+
+			expected_ready := xtradbCluster.Status.HAProxy.Ready
```
[the remark this answers is not in the document]
Comment (BupycHuk) [reply service/k8sclient/k8sclient_test.go]: this is the correct comment

File: service/k8sclient/k8sclient_test.go [hunk 10]
```diff
@@ -75,10 +79,25 @@ func TestK8sClient(t *testing.T) {
...
+			expected_total += xtradbCluster.Status.PMM.Size
+			expected_ready += xtradbCluster.Status.ProxySQL.Ready
+			expected_total += xtradbCluster.Status.ProxySQL.Size
+			expected_ready += xtradbCluster.Status.PXC.Ready
+			expected_total += xtradbCluster.Status.PXC.Size
+
 			cluster, err := getXtraDBCluster(ctx, client, name)
 			require.NoError(t, err)
-			assert.Equal(t, cluster.DetailedState.CountAllPods(), cluster.DetailedState.CountReadyPods())
+			assert.Equal(t, expected_ready, cluster.DetailedState.CountReadyPods())
```
Comment (BupycHuk) [inline service/k8sclient/k8sclient_test.go]: I expected to have just an integer like `2` here and usage `Equal` instead of `Greater`

**Review by BupycHuk — CHANGES_REQUESTED** [review]

File: service/k8sclient/k8sclient_test.go (hunk 10 above)
Comment (jprukner) [reply service/k8sclient/k8sclient_test.go]: I had it like this initially, 2 for xtradb, 9 for mongo. Didn't work for mongo in CI, worked in local test. Let me try it again, maybe I just missed something.

**Comment by BupycHuk** [comment]
as Aleksi already mentioned, please don't squash commits.

**Comment by jprukner** [comment]
> as Aleksi already mentioned, please don't squash commits.

Well, ok. I thought he meant not to squash all commits. I won't do this at all then.

File: service/k8sclient/k8sclient_test.go (hunk 10 above)
Comment (jprukner) [reply service/k8sclient/k8sclient_test.go]: @BupycHuk It seems to be ok now.

**Review by BupycHuk — APPROVED** [review]

**Review by askomorokhov — APPROVED** [review]

**Pull request merged**